### PR TITLE
Artifact Store Apis and Entity commands

### PR DESCRIPTION
**Pull request by akshaydewan**

### CODE REVIEW

File: api/api-artifact-store-config-v1/src/test/groovy/com/thoughtworks/go/apiv1/artifactstoreconfig/ArtifactStoreConfigControllerDelegateTest.groovy [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+          ArtifactStoreRepresenter.toJSON(it, newArtifactStore)
+        }))
+
+        assertThatResponse()
+          .isOk()
+          .hasEtag('"new-md5"')
+          .hasContentType(controller.mimeType)
+          .hasBodyWithJsonObject(newArtifactStore, ArtifactStoreRepresenter)
+      }
+    }
```
Comment (varshavaradarajan) [inline api/api-artifact-store-config-v1/src/test/groovy/com/thoughtworks/go/apiv1/artifactstoreconfig/ArtifactStoreConfigControllerDelegateTest.groovy]: Test for attempted rename of id.

File: server/webapp/WEB-INF/urlrewrite.xml [hunk 2]
```diff
@@ -23,6 +23,18 @@
     http://tuckey.org/urlrewrite/
 -->
 <urlrewrite>
+    <rule>
+        <name>Spark Artifact Store Index API</name>
+        <from>^/api/admin/artifact_stores</from>
```
Comment (varshavaradarajan) [inline server/webapp/WEB-INF/urlrewrite.xml]: Add `(/?)$` at the end.

**Comment by varshavaradarajan** [comment]
The plugin decides whether a property is secure. If there is a secure property, upon create or update, if the plugin is present, the response should contain the encrypted value for the secure property. This does not seem to be happening via the API. Can you please make sure to check for the other spark APIs as well? In rails, encrypted the value of a secure property is done during the time of deserialization - https://github.com/gocd/gocd/blob/master/server/webapp/WEB-INF/rails.new/app/presenters/api_v1/plugin_profile_representer.rb#L37

**Comment by varshavaradarajan** [comment]
Checking the above for all spark APIs is especially important now that we have removed the toggle.

**Comment by varshavaradarajan** [comment]
##### Index

###### Request

```
curl 'http://localhost:8153/go/api/admin/artifact_stores' \
-u 'admin:badger' \
-H 'Accept: application/vnd.go.cd.v1+json'
```

##### Get 

###### Request
```
curl 'http://localhost:8153/go/api/admin/artifact_stores/foobar' \
-u 'admin:badger' \
-H 'Accept: application/vnd.go.cd.v1+json'
```

##### Delete 

###### Request

```
curl 'http://localhost:8153/go/api/admin/artifact_stores' \
-u 'admin:badger' \
-H 'Accept: application/vnd.go.cd.v1+json'
```

##### Create

###### Request

```
curl 'http://localhost:8153/go/api/admin/artifact_stores' \
-u 'admin:badger' \
-H 'Accept: application/vnd.go.cd.v1+json' \
-H 'Content-Type: application/json' \
-X POST -d '{
  "id": "foobar",
  "plugin_id": "cd.go.artifact.docker.registry",
  "properties": [
    {
      "key": "RegistryURL",
      "value": "http://foobar"
    },
    {
      "key": "Username",
      "value": "username"
    },
    {
      "key": "Password",
      "encrypted_value": "5yISOwZeu2RjzVDf7v8HXA=="
    }
  ]
}'
```

##### Update

###### Request 

```
curl 'http://localhost:8153/go/api/admin/artifact_stores/foo' \
-u 'admin:badger' \
-H 'Accept: application/vnd.go.cd.v1+json' \
-H 'Content-Type: application/json' \
-H 'If-Match: "5ee37816f4d807027b762c9f89be7251"' \
-X PUT -d '{
  "id": "foo",
  "plugin_id": "cd.go.artifact.docker.registry",
  "properties": [
   {
      "key": "RegistryURL",
      "value": "http://other-url"
    },
    {
      "key": "Username",
      "value": "username"
    },
    {
      "key": "Password",
      "encrypted_value": "5yISOwZeu2RjzVDf7v8HXA=="
    }
  ]
}' 
```

File: api/api-artifact-store-config-v1/src/main/java/com/thoughtworks/go/apiv1/artifactstoreconfig/representers/ArtifactStoresRepresenter.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+import com.thoughtworks.go.api.base.OutputWriter;
+import com.thoughtworks.go.config.ArtifactStores;
+import com.thoughtworks.go.spark.Routes;
+
+public class ArtifactStoresRepresenter {
+
+    public static void toJSON(OutputWriter writer, ArtifactStores stores) {
+        writer.addLinks(
+                outputLinkWriter -> outputLinkWriter.addLink("self", Routes.ArtifactStoreConfig.BASE))
```
[the remark this answers is not in the document]
Comment (akshaydewan) [reply api/api-artifact-store-config-v1/src/main/java/com/thoughtworks/go/apiv1/artifactstoreconfig/representers/ArtifactStoresRepresenter.java]: Added

File: api/api-artifact-store-config-v1/src/main/java/com/thoughtworks/go/apiv1/artifactstoreconfig/representers/ArtifactStoreRepresenter.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import com.thoughtworks.go.domain.config.ConfigurationProperty;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+public class ArtifactStoreRepresenter {
+    public static void toJSON(OutputWriter outputWriter, ArtifactStore store) {
+        outputWriter
```
[the remark this answers is not in the document]
Comment (akshaydewan) [reply api/api-artifact-store-config-v1/src/main/java/com/thoughtworks/go/apiv1/artifactstoreconfig/representers/ArtifactStoreRepresenter.java]: Added

File: api/api-artifact-store-config-v1/src/main/java/com/thoughtworks/go/apiv1/artifactstoreconfig/ArtifactStoreConfigControllerDelegate.java [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    }
+
+    public String index(Request request, Response response) throws IOException {
+        ArtifactStores artifactStores = artifactStoreService.getPluginProfiles();
+        return writerForTopLevelObject(request, response,
+                outputWriter -> ArtifactStoresRepresenter.toJSON(outputWriter, artifactStores));
+    }
+
+    public String show(Request request, Response response) throws IOException {
+        ArtifactStore artifactStore = getEntityFromConfig(request.params("store_id"));
```
[the remark this answers is not in the document]
Comment (akshaydewan) [reply api/api-artifact-store-config-v1/src/main/java/com/thoughtworks/go/apiv1/artifactstoreconfig/ArtifactStoreConfigControllerDelegate.java]: Changed it to `id`. Makes sense to make it consistent.

File: api/api-artifact-store-config-v1/src/test/groovy/com/thoughtworks/go/apiv1/artifactstoreconfig/ArtifactStoreConfigControllerDelegateTest.groovy (hunk 1 above)
Comment (akshaydewan) [reply api/api-artifact-store-config-v1/src/test/groovy/com/thoughtworks/go/apiv1/artifactstoreconfig/ArtifactStoreConfigControllerDelegateTest.groovy]: Done

File: server/webapp/WEB-INF/urlrewrite.xml (hunk 2 above)
Comment (akshaydewan) [reply server/webapp/WEB-INF/urlrewrite.xml]: Made this consistent across all Spark APIs and SPAs

**Comment by akshaydewan** [comment]
Added fix for encrypting values while deserializing.

**Pull request merged**